doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+on which hints are defined and the documentation for each specific <code
+class="docutils literal notranslate"><span
+class="pre">expand_*hint*()</span></code> function (e.g.,
+{func}`~.expand_trig`) for details on what each hint is designed to do.
+
+The `**hints` keyword arguments are additional hints that may be passed to the
+expand function to specify additional behavior (these are separate from the
+predefined *hints* described in the previous paragraph). Unknown hints should
+be ignored as they may apply to other functions' custom `expand()` methods. A
+common hint to define is `force`, where `force=True` would force an expansion
+that might not be mathematically valid for all the given input assumptions.
+For example, `expand_log(log(x*y), force=True)` produces `log(x) + log(y)`
+even though this identity is not true for all complex `x` and `y` (typically
+`force=False` is the default).
+
+Note that `expand()` automatically takes care of recursively expanding
+expressions using its own `deep` flag, so `_eval_expand_*` methods should not
+recursively call expand on the arguments of the function.
+
+For our [`versin` example](custom-functions-versine-definition), we can define
+rudimentary `trig` expansion by defining an `_eval_expand_trig` method,
+which recursively calls `expand_trig()` on `1 - cos(x)`:
+
+```
+>>> from sympy import expand_trig
+>>> y = symbols('y')
+>>> class versin(Function):
+...    def _eval_expand_trig(self, **hints):
+...        x = self.args[0]
+...        return expand_trig(1 - cos(x))
+>>> versin(x + y).expand(trig=True)
+sin(x)*sin(y) - cos(x)*cos(y) + 1
+```
+
+A more sophisticated implementation might attempt to rewrite the result of
+`expand_trig(1 - cos(x))` back into `versin` functions. This is left as an
+exercise for the reader.
+
+(custom-functions-differentiation)=
+### Differentiation
+
+To define differentiation via {func}`~.diff`, define a method `fdiff(self,
+argindex)`. `fdiff()` should return the derivative of the function, without
+considering the chain rule, with respect to the `argindex`-th variable.
+`argindex` is indexed starting at `1`.
+
+That is, `f(x1, ..., xi, ..., xn).fdiff(i)` should return $\frac{d}{d x_i}
+f(x_1, \ldots, x_i, \ldots, x_n)$, where $x_k$ are independent of one another.
+`diff()` will automatically apply the chain rule using the result of
+`fdiff()`. User code should use `diff()` and not call `fdiff()` directly.
+
+```{note}
+
+`Function` subclasses should define differentiation using `fdiff()`. Subclasses
+of {class}`~.Expr` that aren't `Function` subclasses will need to define
+`_eval_derivative()` instead. It is not recommended to redefine
+`_eval_derivative()` on a `Function` subclass.
+
+```
+
+For our [$\operatorname{versin}$ example
+function](custom-functions-versine-definition), the derivative is $\sin(x)$.
+
+```py
+>>> class versin(Function):
+...     def fdiff(self, argindex=1):
+...         return sin(self.args[0])
```

Comment:
It's mentioned in the text above but I'll also note it here with a comment. I decided to ignore invalid `argindex`, since that should never happen from `diff` and users shouldn't really ever call `fdiff` (see https://github.com/sympy/sympy/issues/23490).